internal/gps/project_root_cache.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import "github.com/pkg/errors"
+
+// ProjectRootCache caches the result of deducing a project root from a
+// source manager, which is quite helpful when repeatedly converting from
+// import paths to root projects.
+type ProjectRootCache struct {
+	prs map[string]ProjectRoot
```

Comment:
that logic is definitely complex, because it has to be concurrency-safe, but yes, the `rootxt` radix tree is the cache, and [this is basically the cache "get"](https://github.com/golang/dep/blob/7d365252964cbb738685d2ed2ae26f208e111424/internal/gps/deduce.go#L585).

we use the radix tree, rather than a map, so that we only ever have to actually deduce a project root once, even if we haven't seen a literal path before. so, the sequence goes something like this:

1. a call comes in to deduce `k8s.io/kubernetes/pkg/foo`
2. we make the necessary HTTP call and find that `k8s.io/kubernetes` is the root, and record that in the radix tree
3. a subsequent call comes in to deduce `k8s.io/kubernetes/pkg/bar`

even though no exact call happened for `k8s.io/kubernetes/pkg/bar` before, or even for its actual root, `k8s.io/kubernetes`, this call still results in a cache hit.

this is one of the things afforded to us by the assumption that projects are repo root-only, and therefore necessarily cannot be nested.